### ISSUE TITLE
Bug in Cipher Feedback (CFB) encryption/decryption (tracking from pycrypto)

### ISSUE THREAD

**Issue by MartyMacGyver** [Legrandin/pycryptodome#78]
There are reports of a probable bug in Cipher Feedback (CFB) mode that were raised in https://github.com/dlitz/pycrypto:

https://github.com/dlitz/pycrypto/issues/226
https://github.com/dlitz/pycrypto/issues/122

A demonstration of the bug (in Python 2.7) is presented here:
https://github.com/dlitz/pycrypto/issues/226#issuecomment-311085292

I'm able to reproduce this output with pycryptodome 3.4.6.

In summary, CFB mode encryption/decryption seems to produce incorrect values for NIST test cases from http://csrc.nist.gov/groups/STM/cavp/block-ciphers.html#aes:
http://csrc.nist.gov/groups/STM/cavp/documents/aes/KAT_AES.zip

**Comment by Legrandin** [comment]
Thanks for flagging this, but would you mind indicating which NIST test fails exactly?
PyCryptodome already includes them in the test suite and they all seem to pass.

Also, are you sure this is not about the API?

`cryptography.hazmat.primitives.ciphers.modes.CFB` implicitly sets *segment size* to **128 bits** (though the [`cryptography` API docs](https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption) are silent about this).

On the other hand, `Crypto.Cipher.AES` in CFB mode accepts *segment size* as a parameter (`segment_size`) with default value of **8 bits**. See [its API](http://legrandin.github.io/pycryptodome/Doc/3.4.6/Crypto.Cipher.AES-module.html).

In other words, the snippet in https://github.com/dlitz/pycrypto/issues/226#issuecomment-311085292 is comparing two different ciphers. 

Finally, the [buffer overrun bug](https://github.com/dlitz/pycrypto/issues/122) is only affecting PyCrypto and not PyCryptodome: the CFB implementation in PyCryptodome has been fully rewritten (it is arguably a tire fire in PyCrypto).

**Comment by MartyMacGyver** [comment]
First, thanks for your work on keeping this important package alive and growing! I refer people to this project when pycrypto questions come up that are easily solved by simply moving to pycryptodome.

In this case I was passing this along as I noticed this was re-reported relatively recently as an issue. It seemed reproducible, but you're right that if it's not a fair data comparison then it may well not be a bug.

**Comment by vog** [comment]
@Legrandin wrote:

> Also, are you sure this is not about the API?

Maybe, but then it's a issue about strange defaults, I guess.

**Comment by Legrandin** [comment]
I do agree on the importance on clear, documented, unsurprising crypto APIs.

This default was decided long time ago in PyCrypto and not breaking existing applications is a priority, hence it stays (*segment size* also not easily being the cause of any security issue; I did break the API for security hazards like raw RSA). Having said that, *segment size* is an important CFB parameter and one might argue that defaulting to CFB-8 is equally good to defaulting to CFB-128: the API documents that the choice done anyway.

I close this issue, since it seems to require no action.
